lib/handler/connect.c

```diff
@@ -0,0 +1,189 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+#include "h2o/hostinfo.h"
+#include "h2o/memory.h"
+#include "h2o/socket.h"
+#include "h2o.h"
+
+struct st_connect_handler_t {
+    h2o_handler_t super;
+    h2o_proxy_config_vars_t config;
+};
+
+#define MAX_CONNECT_RETRIES 3
+
+struct st_server_address_t {
+    struct sockaddr *sa;
+    socklen_t salen;
+};
+
+struct st_connect_request_t {
+    struct st_connect_handler_t *handler;
+    h2o_loop_t *loop;
+    h2o_req_t *src_req;
+    h2o_socket_t *sock;
+    h2o_hostinfo_getaddr_req_t *getaddr_req;
+    struct {
+        h2o_iovec_t host;
+        char port[sizeof(H2O_UINT16_LONGEST_STR)];
+    } server_name;
+    struct {
+        struct st_server_address_t list[MAX_CONNECT_RETRIES];
+        size_t size;
+        size_t next;
+    } server_addresses;
+    h2o_timer_t timeout;
+};
+
+static void start_connect(struct st_connect_request_t *creq);
+
+static void on_error(struct st_connect_request_t *creq, const char *errstr)
+{
+    h2o_timer_unlink(&creq->timeout);
+    h2o_send_error_502(creq->src_req, "Gateway Error", errstr, 0);
+}
+
+static void on_timeout(h2o_timer_t *entry)
+{
+    struct st_connect_request_t *creq = H2O_STRUCT_FROM_MEMBER(struct st_connect_request_t, timeout, entry);
+    on_error(creq, h2o_httpclient_error_io_timeout);
+}
+
+static void on_connect(h2o_socket_t *sock, const char *err)
+{
+    struct st_connect_request_t *creq = sock->data;
+
+    if (err) {
+        if (creq->server_addresses.next == creq->server_addresses.size) {
+            on_error(creq, err);
+            return;
+        }
+        start_connect(creq);
+        return;
+    }
+
+    h2o_timer_unlink(&creq->timeout);
+    h2o_req_t *req = creq->src_req;
+    uint64_t timeout = creq->handler->config.tunnel.timeout;
+    sock->data = NULL;
+    creq->sock = NULL;
+    req->res.status = 200;
+
+    h2o_tunnel_t *tunnel = h2o_tunnel_create_from_socket(sock);
+    req->establish_tunnel(req, tunnel, timeout);
+}
+
+static void on_generator_dispose(void *_self)
+{
+    struct st_connect_request_t *creq = _self;
+
+    if (creq->getaddr_req != NULL) {
+        h2o_hostinfo_getaddr_cancel(creq->getaddr_req);
+        creq->getaddr_req = NULL;
+    }
+
+    if (creq->sock != NULL)
+        h2o_socket_close(creq->sock);
+}
+
+static void store_server_addresses(struct st_connect_request_t *creq, struct addrinfo *res)
```

Comment:
Thinking about v4/v6 selection, I think we would probably be good enough to simply order v6 ahead of v4 addresses. As long as there are fewer than 3 v6 addresses, we'll try a v4 address too. Alternatively, you could interleave v6 and v4, with v6 leading. WDYT?

Reply:
For the OSS branch, I do not think there is a strategy that is better than deferring the problem to `getaddrinfo`.